particle-iot/device-os
# Fix/gen3 spi concurrent

### Problem

The EthernetWing uses the high speed SPI for Gen3 devices, but the hardware design blocks to use the 2nd SPI interface unless to do some modifications to the EthernetWing.

Besides the above, the high speed SPI was not able to be shared with other SPI peripheral devices to use it concurrently with the EthernetWing since the SPI HAL has not implemented any synchronization mechanism.

### Solution

Adds HAL_SPI_Acquire() and HAL_SPI_Release() in all the functions of SPI HAL and exposes these two to Wiring layer.

### Steps to Test

Tested on a Xenon + EthernetWing + ST7735 SPI TFT Display.

Use the following Example App to test, the TFT Display and the Cloud connection should work as expected to allow user apps to use.

### Example App

```c
#include <Adafruit_ST7735.h>

STARTUP(System.enableFeature(FEATURE_ETHERNET_DETECTION));
SYSTEM_MODE(MANUAL);

#define CS   A1
#define DC   A0
#define RST  A2  

Adafruit_ST7735 tft = Adafruit_ST7735(CS, DC, RST);

void setup() {
    Serial.begin(115200);
    delay(3000);
    Serial.println("SPI Test Started");

    tft.initG();

    Ethernet.on();
    Ethernet.connect();
    Serial.println("Waiting to connect to Ethernet...");
    while(!Ethernet.ready());
    Serial.println("Connected to Ethernet.");
    
    Particle.connect();
}

void loop() {
    SPI.lock();
    tft.fillScreen(ST7735_BLACK);
    tft.setCursor(0, 0);
    tft.setTextColor(ST7735_WHITE);
    tft.setTextWrap(true);
    tft.print("HELLO!");
    SPI.unlock();
    delay(300);

    SPI.lock();
    tft.fillScreen(ST7735_BLACK);
    tft.setCursor(0, 0);
    tft.setTextColor(ST7735_WHITE);
    tft.setTextWrap(true);
    tft.print("PARTICLE");    
    SPI.unlock(); 
    delay(300);
}
```

### References

[CH35954]

---

### Completeness
